src/bun.js/bindings/NodeVM.cpp

```diff
@@ -67,138 +76,478 @@ JSC::EncodedJSValue INVALID_ARG_VALUE_VM_VARIATION(JSC::ThrowScope& throwScope,
-            return nullptr;
-        return WebCore::subspaceForImpl<NodeVMScript, WebCore::UseCustomHeapCellType::No>(
-            vm,
-            [](auto& spaces) { return spaces.m_clientSubspaceForNodeVMScript.get(); },
-            [](auto& spaces, auto&& space) { spaces.m_clientSubspaceForNodeVMScript = std::forward<decltype(space)>(space); },
-            [](auto& spaces) { return spaces.m_subspaceForNodeVMScript.get(); },
-            [](auto& spaces, auto&& space) { spaces.m_subspaceForNodeVMScript = std::forward<decltype(space)>(space); });
+        JSValue produceCachedDataOpt = options->getIfPropertyExists(globalObject, Identifier::fromString(vm, "produceCachedData"_s));
+        if (produceCachedDataOpt && !produceCachedDataOpt.isUndefined()) {
+            RETURN_IF_EXCEPTION(scope, {});
+            if (!produceCachedDataOpt.isBoolean()) {
+                ERR::INVALID_ARG_TYPE(scope, globalObject, "options.produceCachedData"_s, "boolean"_s, produceCachedDataOpt);
+                return false;
+            }
+            outProduceCachedData = produceCachedDataOpt.asBoolean();
+            return true;
+        }
+        return false;
     }
 
-    static void destroy(JSC::JSCell*);
-    static JSC::Structure* createStructure(JSC::VM& vm, JSC::JSGlobalObject* globalObject, JSC::JSValue prototype)
+    bool validateCachedData(JSC::JSGlobalObject* globalObject, JSC::VM& vm, JSC::ThrowScope& scope, JSObject* options, std::vector<uint8_t>& outCachedData)
     {
-        return JSC::Structure::create(vm, globalObject, prototype, JSC::TypeInfo(JSC::ObjectType, StructureFlags), info());
-    }
+        JSValue cachedDataOpt = options->getIfPropertyExists(globalObject, Identifier::fromString(vm, "cachedData"_s));
+        if (cachedDataOpt && !cachedDataOpt.isUndefined()) {
+            RETURN_IF_EXCEPTION(scope, {});
+            if (!cachedDataOpt.isCell()) {
+                ERR::INVALID_ARG_INSTANCE(scope, globalObject, "options.cachedData"_s, "Buffer, TypedArray, or DataView"_s, cachedDataOpt);
+                return false;
+            }
 
-    static JSObject* createPrototype(VM& vm, JSGlobalObject* globalObject);
+            // If it's a cell, verify it's a Buffer, TypedArray or DataView
+            if (cachedDataOpt.isCell()) {
+                bool isValidType = false;
 
-    const JSC::SourceCode& source() const { return m_source; }
+                if (auto* arrayBufferView = JSC::jsDynamicCast<JSC::JSArrayBufferView*>(cachedDataOpt)) {
+                    if (!arrayBufferView->isDetached()) {
+                        std::span<const uint8_t> span = arrayBufferView->span();
+                        outCachedData = { span.begin(), span.end() };
+                        isValidType = true;
+                    }
+                } else if (auto* arrayBuffer = JSC::jsDynamicCast<JSC::JSArrayBuffer*>(cachedDataOpt)) {
+                    std::span<const uint8_t> span = arrayBuffer->impl()->span();
```

Comment:
ArrayBuffer can also be detached or an empty buffer, in which case this will segfault deu to the null pointer dereference.

This needs a check if the impl is null